src/Agent.Listener/CapabilitiesProvider.cs

```diff
@@ -0,0 +1,291 @@
+        public string[] Paths { get; private set; }
+    }
+
+    public class ToolCapability
+    {
+        public ToolCapability(string name, string command, string commandArgs = null)
+        {
+            Name = name;
+            Command = command;
+            CommandArgs = commandArgs;
+        }
+
+        //Name of the capability as requried by the task definition.
+        public string Name { get; private set; }
+
+        //Name of an executable file, which can determine if a capability is present.
+        //The script is expected to print on the standard output the value for the capability.
+        public string Command { get; private set; }
+
+        //Arguments passed to Command.
+        public string CommandArgs { get; private set; }
+    }
+
+    public sealed class CapabilitiesProvider : AgentService, ICapabilitiesProvider
+    {
+        private List<Capability> _regularCapabilities =
+            new List<Capability>
+            {
+                new Capability( "ant" ),
+                new Capability( "bundler", "bundle" ),
+                new Capability( "clang" ),
+                new Capability( "cmake" ),
+                new Capability( "curl" ),
+                new Capability( "git" ),
+                new Capability( "gulp" ),
+                new Capability( "java" ),
+                new Capability( "JDK", "javac" ),
+                new Capability( "make" ),
+                new Capability( "maven", "mvn" ),
+                new Capability( "MSBuild", "xbuild" ),
+                new Capability( "node.js", "node" ),
+                new Capability( "node.js", "nodejs" ),
+                new Capability( "npm" ),
+                new Capability( "python" ),
+                new Capability( "python3" ),
+                new Capability( "sh" ),
+                new Capability( "subversion", "svn" ),
+                new Capability( "ruby" ),
+                new Capability( "rake" ),
+                new Capability( "Xamarin.iOS", "mdtool", new string[] { "/Applications/Xamarin Studio.app/Contents/MacOS/mdtool" } ),
+                new Capability( "Xamarin.Android", "mandroid", new string[] { "/Library/Frameworks/Xamarin.Android.framework/Commands/mandroid" } )
+            };
+
+        private static readonly List<ToolCapability> _toolCapabilities =
+            new List<ToolCapability>
+            {
+                new ToolCapability( "xcode", "xcode-select", "-p" ),
+            };
+
+        private static readonly string[] _ignoredEnvVariables = new string[] {
+            "TERM_PROGRAM",
+            "TERM",
+            "TERM_PROGRAM_VERSION",
+            "SHLVL",
+            "ls_colors",
+            "comp_wordbreaks"
+        };
+
+        // Ignore env vars specified in the 'VSO_AGENT_IGNORE' env var
+        private const string EnvIgnore = "VSO_AGENT_IGNORE";
+
+        private Dictionary<string, string> _capsCache;
```

Comment:
It will give me a compiler error if I make it readonly. I initialize the cache only after GetRegularCapabilities has been invoked.